### PR TITLE
feat(docker): add Dockerfile for trame apps

### PR DESCRIPTION
**Pull request by psavery**
The trame apps will be able to mount a special "deploy" directory into the container and start a
service for multiple clients to connect to and individually run the trame app.

### CODE REVIEW

File: docker/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,77 @@
+FROM ubuntu:20.04
```
Comment (jourdain) [inline docker/Dockerfile:1]: We want it to be an ARG so we can swap between std ubuntu or ubuntu-nivida for access to GPU.

```
ARG BASE_IMAGE=ubuntu:20.04
FROM ${BASE_IMAGE}
```

File: docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+      apache2-utils \
+      sudo && \
+    rm -rf /var/lib/apt/lists/*
+
+# Set python3 to python3.9 (otherwise, it will be python3.8)
+RUN update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.9 1
+
+# Install and upgrade pip
+RUN wget -O- https://bootstrap.pypa.io/get-pip.py | python3.9 && \
+    pip install -U pip
```
Comment (jourdain) [inline docker/Dockerfile:31]: Should we use some extra args to remove cache files since we are inside docker?

File: docker/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+# Install setup dependencies
+RUN pip install PyYAML
```
Comment (jourdain) [inline docker/Dockerfile:34]: same here, remove pip cache

File: docker/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+# Set up needed permissions and users
+RUN groupadd proxy-mapping && \
+    groupadd trame-user && \
+    useradd --system -g trame-user -G proxy-mapping -s /sbin/nologin trame-user && \
+    usermod -a -G proxy-mapping www-data && \
+    useradd admin && echo "admin:admin" | chpasswd && adduser admin sudo && \
+    mkdir -p /opt/trame/log && \
+    chown -R trame-user:trame-user /opt/trame && \
```
Comment (jourdain) [inline docker/Dockerfile:43]: wondering is we need `mkdir -p /deploy/server` so it can work when we do `-v /.../:/deploy/server`.

**Comment by jourdain** [comment]
Once ready we should try to publish those images via ci or else?

```
kitware/trame:ubuntu-20.04-py39 (BASE_IMAGE=ubuntu:20.04)
kitware/trame:ubuntu-20.04-py39-nvidia-460.73.01 (BASE_IMAGE=nvidia/driver:460.73.01-ubuntu20.04)
```

File: docker/Dockerfile (hunk 1 above)
Comment (psavery) [reply docker/Dockerfile]: Fixed

File: docker/Dockerfile (hunk 2 above)
Comment (psavery) [reply docker/Dockerfile]: Looks like adding `--no-cache-dir` keeps it from caching files. I added that.

File: docker/Dockerfile (hunk 3 above)
Comment (psavery) [reply docker/Dockerfile]: Added `--no-cache-dir`

File: docker/Dockerfile (hunk 4 above)
Comment (psavery) [reply docker/Dockerfile:44]: I gave the mounting a try without the `mkdir -p /deploy/server`, and it seems to work. We might be able to skip this.

**Comment by psavery** [comment]
> Once ready we should try to publish those images via ci or else?
> 
> ```
> kitware/trame:ubuntu-20.04-py39 (BASE_IMAGE=ubuntu:20.04)
> kitware/trame:ubuntu-20.04-py39-nvidia-460.73.01 (BASE_IMAGE=nvidia/driver:460.73.01-ubuntu20.04)
> ```

Sure! I think CI is a good idea.

We can create images with those tags. Maybe we can also make the default `kitware/trame` point to ` kitware/trame:ubuntu-20.04-py39` (so the user can just run `docker run kitware/trame` without specifying the tag), and maybe `kitware/trame:nvidia` can point to the latest nvidia build (currently `kitware/trame:ubuntu-20.04-py39-nvidia-460.73.01`)?

File: docker/Dockerfile (hunk 4 above)
Comment (jourdain) [reply docker/Dockerfile:44]: yes I was not sure since I did not try

File: docker/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+# Install setup dependencies
+RUN pip install PyYAML
+
+# Set up needed permissions and users
+RUN groupadd trame-user -g 1000 && \
+    groupadd proxy-mapping -g 1001 && \
+    useradd -u 1000 -g trame-user -G proxy-mapping -s /sbin/nologin trame-user && \
+    usermod -a -G proxy-mapping www-data && \
+    mkdir -p /opt/trame/log && \
```
Comment (psavery) [inline docker/Dockerfile:46]: Do you know if this log file is being used for anything? Or can we delete it?

File: docker/scripts/fix_permissions.sh [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  run_chown=true
+fi
+
+if [[ "$run_chown" == false ]]; then
+  exit 0
+fi
+
+# Run chown on all trame-user directories/files
+chown -R trame-user:trame-user /opt/trame
+chown trame-user:proxy-mapping /opt/trame/proxy-mapping.txt
```
Comment (psavery) [inline docker/scripts/fix_permissions.sh:26]: @jourdain This file modifies the `trame-user` UID and GID to match the user's UID and GID if they don't match. This fixes the Linux permissions issue involving UID/GID (it works in my testing).

File: docker/Dockerfile (hunk 5 above)
Comment (jourdain) [reply docker/Dockerfile:46]: no as we use the /deploy/server/logs instead

File: docker/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    a2enmod proxy_http && \
+    a2enmod proxy_wstunnel && \
+    a2enmod rewrite && \
+    a2enmod headers && \
+    a2dissite 000-default.conf && \
+    a2ensite 001-trame && \
+    a2dismod autoindex -f
+
+# Open port 80 to the world outside the container
+EXPOSE 80
```
Comment (jourdain) [inline docker/Dockerfile:79]: When you tried 8080 in the vhost, did you update that value as well?

**Comment by jourdain** [comment]
LGTM but lets rewrite the commits to not trigger a new release since the Python code did not change.

`chore(docker): add container for cloud deployment`

File: docker/Dockerfile (hunk 7 above)
Comment (psavery) [reply docker/Dockerfile:79]: Yes I did. I'm happy to try other things too if you'd like.

File: docker/Dockerfile (hunk 5 above)
Comment (psavery) [reply docker/Dockerfile]: Fixed.

**Pull request merged**